piet-cairo/src/text.rs

```diff
@@ -210,233 +396,198 @@ impl TextLayout for CairoTextLayout {
-        }
-
-        let height = self
-            .line_metrics
-            .last()
-            .map(|lm| lm.y_offset + lm.height)
-            .unwrap_or(0.0);
-
-        // determine whether this click is within the y bounds of the layout,
-        // and what line it coorresponds to. (For points above and below the layout,
-        // we hittest the first and last lines respectively.)
-        let (y_inside, lm) = if point.y < 0. {
-            (false, self.line_metrics.first().unwrap())
-        } else if point.y >= height {
-            (false, self.line_metrics.last().unwrap())
+        let (is_inside, index, trailing) = self.pango_layout.xy_to_index(x as i32, y as i32);
+        let index = if trailing == 0 {
+            index.try_into().unwrap()
         } else {
-            let line = self
-                .line_metrics
-                .iter()
-                .find(|l| point.y >= l.y_offset && point.y < l.y_offset + l.height)
-                .unwrap();
-            (true, line)
+            /*
+             * NOTE(ForLoveOfCats): The docs specify that a non-zero value for trailing
+             * indicates that the point aligns to the trailing edge of the grapheme. In
+             * that case the value tells us the number of "characters" in the grapheme.
+             */
+
+            let text = self.pango_layout.get_text().unwrap();
+            let index = index.try_into().unwrap();
+            let trailing = trailing.try_into().unwrap();
+
+            text[index..]
+                .char_indices()
+                .nth(trailing)
+                .map(|(offset, _)| index + offset)
+                .unwrap_or_else(|| text.len())
         };
 
-        // Trailing whitespace is remove for the line
-        let line = &self.text[lm.range()];
-
-        let mut htp = hit_test_line_point(&self.font, line, point);
-        htp.idx += lm.start_offset;
-        if htp.idx == lm.end_offset {
-            htp.idx -= util::trailing_nlf(line).unwrap_or(0);
-        }
-        htp.is_inside &= y_inside;
-        htp
-    }
-
-    fn hit_test_text_position(&self, idx: usize) -> HitTestPosition {
-        let idx = idx.min(self.text.len());
-        assert!(self.text.is_char_boundary(idx));
-
-        if idx == 0 && self.text.is_empty() {
-            return HitTestPosition::new(Point::new(0., self.font.extents().ascent), 0);
-        }
-
-        // first need to find line it's on, and get line start offset
-        let line_num = util::line_number_for_position(&self.line_metrics, idx);
-        let lm = self.line_metrics.get(line_num).cloned().unwrap();
-
-        let y_pos = lm.y_offset + lm.baseline;
+        let (metric_index, metric) = self
+            .line_metrics
+            .iter()
+            .enumerate()
+            .find(|(_, metric)| {
+                metric.start_offset <= index
+                    && index <= metric.end_offset - metric.trailing_whitespace
+            })
+            .unwrap();
```

Comment:
it looks like this will fail if the text position is in the trailing whitespace, which is probably possible? In the coretext backend we have a helper fn for this:

```rust
fn line_number_for_utf8_offset(&self, offset: usize) -> usize {
        match self
            .line_metrics
            .binary_search_by_key(&offset, |lm| lm.start_offset)
        {
            Ok(line) => line,
            Err(line) => line.saturating_sub(1),
        }
    }

```

Reply:
With the way the values we get are this shouldn't be a problem by my understanding. Using the text from `eol_hit_testing` "AA AA\nAA" 
This produces three line metrics with start ends [(0, 3), (3, 6), (6, 8)]
The end of the first metric is at
```
012345678
AA AA AA
   ^
```
So once the trailing whitespace is subtracted you get
```
012345678
AA AA AA
  ^
```
The same applies for the end of the second metric. The third metric however has no trailing whitespace so it produces the expected end at
```
012345678
AA AA AA
        ^
```
Let me know if I am massively misinterpreting this

Reply:
Yeah this was wrong for a multitude of reasons. I've improved the logic which also happens to have fixed a bug I didn't even know existed
Before
![image](https://user-images.githubusercontent.com/30666851/111020002-d9f25a00-8390-11eb-9032-86b7889a1fc0.png)
After
![image](https://user-images.githubusercontent.com/30666851/111020011-e8407600-8390-11eb-9053-c8074abc8f48.png)